src/efftester.ml

```diff
@@ -1489,4 +1525,19 @@ let eq_test =
       | Some t -> nativeByteEquivalence (toOCaml (print_wrap t)))
 ;;
 
+let rand_eq_test =
+  let typ = Gen.generate1 (Gen.oneofl [ Int; Float; String ]) in
```

Comment:
I agree that for one top-level invocation, all tests will have the same `typ`.

However I would recommend against this style for reproducability.
In case the generator actually triggers a bug we would want to be able to reproduce the problematic run
with a given seed. You will not immediately be able to so if you mix in`Gen.generate1`, as it kicks off its own random number generator: https://github.com/c-cube/qcheck/blob/master/src/core/QCheck.ml#L255

So what do we do? A simple solution if we want to run separate `int`, `float`, and `string` tests is to write a parameterized test and just invoke it three times.
A more complex solution is to use the `Gen` module's bind operator `>>=` to write a generator of a dependent pair `(typ, term)` such that the type of `term` is `typ`. 

http://janmidtgaard.dk/quickcheck/slides/lec06.pdf  (p.24 - called 20/35)
contains a simple example of such a dependent generator (of numbers).
Here's some code realizing something along the above (with an option type on top to represent failure of the generator's backtracking):
```
let gen =
    Gen.(basetypeGen >>= fun typ ->
         termGen initTriEnv typ >>= fun t_opt -> match t_opt with
         | None -> return None
         | Some t -> return (Some (typ,t)))
```